neo/Wallets/Wallet.cs

```diff
@@ -182,77 +214,142 @@ public Transaction MakeTransaction(IEnumerable<TransactionAttribute> attributes,
                 if (!Contains(from)) return null;
                 accounts = new[] { from };
             }
-            TransactionAttribute[] attr = attributes?.ToArray() ?? new TransactionAttribute[0];
-            using (Snapshot snapshot = Blockchain.Singleton.GetSnapshot())
+
+            HashSet<UInt160> cosigners = new HashSet<UInt160>();
+            byte[] script;
+            List<(UInt160 Account, BigInteger Value)> balances_gas = null;
+            using (ScriptBuilder sb = new ScriptBuilder())
             {
-                uint validUntilBlock = snapshot.Height + Transaction.MaxValidUntilBlockIncrement;
-                foreach (UInt160 account in accounts)
+                foreach (var (assetId, group, sum) in outputs.GroupBy(p => p.AssetId, (k, g) => (k, g, g.Select(p => p.Value.Value).Sum())))
                 {
-                    Transaction tx = MakeTransaction(snapshot, 0, nonce, totalPay, outputs, account, validUntilBlock, attr);
-                    if (tx != null) return tx;
+                    var balances = new List<(UInt160 Account, BigInteger Value)>();
+                    foreach (UInt160 account in accounts)
+                        using (ScriptBuilder sb2 = new ScriptBuilder())
+                        {
+                            sb2.EmitAppCall(assetId, "balanceOf", account);
+                            ApplicationEngine engine = ApplicationEngine.Run(sb2.ToArray(), snapshot, testMode: true);
+                            if (engine.State.HasFlag(VMState.FAULT)) return null;
+                            BigInteger value = engine.ResultStack.Pop().GetBigInteger();
+                            if (value.Sign > 0) balances.Add((account, value));
+                        }
+                    BigInteger sum_balance = balances.Select(p => p.Value).Sum();
+                    if (sum_balance < sum) return null;
+                    foreach (TransferOutput output in group)
+                    {
+                        balances = balances.OrderBy(p => p.Value).ToList();
+                        var balances_used = FindPayingAccounts(balances, output.Value.Value);
+                        cosigners.UnionWith(balances_used.Select(p => p.Account));
+                        foreach (var (account, value) in balances_used)
+                        {
+                            if (account != output.ScriptHash)
```

Comment:
We still need to support sending tokens to the sender himself. Some application may use it. For example: claiming GAS.

Reply:
Reverted

Reply:
Claiming GAS is now one step, right? Just the sending to yourself part.